prboom2/src/p_maputl.c

```diff
@@ -725,6 +725,195 @@ dboolean P_PathTraverse(fixed_t x1, fixed_t y1, fixed_t x2, fixed_t y2,
   return P_TraverseIntercepts(trav, FRACUNIT);
 }
 
+//
+// RoughBlockCheck
+// [XA] adapted from Hexen -- used by P_RoughMonsterSearch
+//
+
+static mobj_t *RoughBlockCheck(mobj_t *mo, int index, angle_t fov)
+{
+  mobj_t *link;
+  mobj_t *master;
+  angle_t angle, minang, maxang;
+
+  // pre-calculate fov check stuff since it
+  // won't change during the blocklinks loop
+  if (fov > 0) {
+    minang = mo->angle - fov / 2;
+    maxang = mo->angle + fov / 2;
+  }
+
+  link = blocklinks[index];
+  while (link)
+  {
+    // skip non-shootable actors
+    if (!(link->flags & MF_SHOOTABLE))
+    {
+      link = link->bnext;
+      continue;
+    }
+
+    // skip other players in co-op
+    if (netgame && !deathmatch && link->player)
+    {
+      link = link->bnext;
+      continue;
+    }
+
+    // skip the projectile's owner
+    if (link == mo->target)
+    {
+      link = link->bnext;
+      continue;
+    }
+
+    // skip friendly actors if owner is friendly (or a player)
+    if ((link->flags & MF_FRIEND) && mo->target != NULL &&
```

Comment:
Players have MF_FRIEND set. I'd say we shouldn't have an extra player check unless we need it. Someone may want to have a mechanism that shuts off this flag on the player, and then they would want this condition to change as well.

`if (mo->target && (link->flags & mo->target->flags & MF_FRIEND))`

If this should also exclude non-friends targeting non-friends:

`if (mo->target && !((link->flags ^ mo->target->flags) & MF_FRIEND)))`

Reply:
Thanks! That second one ought to make this function work nicely for monsters, which is kinda niche but someone somewhere will find a use for it I'm sure. :P

I think the first one could handle the deathmatch player case by adding in `&& (!deathmatch || !link->player)` at the end. Presuming my brain's workin' right ATM (never a guarantee with me :P ).